### PR TITLE
GPU Acceleration Addition

### PR DESCRIPTION
**Pull request by JMCTek**
I have added GPU acceleration by changing the containerd .toml and deploying a runtime class within kubernetes.

This configuration can be set by the USE_GPU_ACCEL=true argument.

Nvidia GPU is accessible immediately after configuration in container environments.

### CODE REVIEW

File: templates/00_tn_header.sh [hunk 1]
```diff
@@ -14,4 +14,5 @@ set -o noglob
 #   INSECURE=true (default: false, for OS builds, or if you want to run without ssl cert verification)
 #   INSTALL_WATCHDOG=false (default: true, enable or disable the teknoir watchdog service)
 #   USE_DOCKER=false (default: true, enable or disable the use of docker as container runtime for k3s)
+#   USE_GPU_ACCEL=false (default: true, enable or disable the use of GPU acceleration in containerd)
```
Comment (tekn0ir) [inline templates/00_tn_header.sh]: Nowhere in this script is USE_GPU_ACCEL set to true by default.

File: templates/14_tn_configure_k3s.sh [hunk 2]
```diff
@@ -0,0 +1,61 @@
+info "Configure Rancher K3s"
+
+if [ "${USE_GPU_ACCEL}" = true ]; then
+    info "Use nvidia container runtime in containerd"
+    config_file="/var/lib/rancher/k3s/agent/etc/containerd/config.toml.tmpl"
```
Comment (tekn0ir) [inline templates/14_tn_configure_k3s.sh]: Adding a config.toml.tmpl is not needed according to K3s docs:
https://docs.k3s.io/advanced#nvidia-container-runtime-support

Also, this causes all PODs not to be scheduled after rebooting the system the first time. It does not recover until /var/lib/rancher/k3s/agent/etc/containerd/config.toml.tmpl is deleted.

File: templates/14_tn_configure_k3s.sh [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    nvidia_runtimeclass_file="$nvidia_runtimeclass_dir/nvidia-runtimeclass.yaml"
+
+    # Check if the file already exists
+    if [ -e "$nvidia_runtimeclass_file" ]; then
+        echo "File already exists at $nvidia_runtimeclass_file. Deleting and recreating."
+        $SUDO rm -f "$nvidia_runtimeclass_file"
+    fi
+
+    # Create Kubernetes runtime class yaml with the specified contents
+    $SUDO tee "$nvidia_runtimeclass_file" > /dev/null << EOL
```
Comment (tekn0ir) [inline templates/14_tn_configure_k3s.sh]: I think the RuntimeClass can be added to /var/lib/rancher/k3s/server/manifests/teknoir.yaml instead. That ensures that it is created persistently.

**Review by tekn0ir — CHANGES_REQUESTED** [review]
I cannot get HW acceleration to work, even if I fiddle with it. How did you test it?

**Comment by tekn0ir** [comment]
Ok, after deleting the config.toml.tmpl file I was able to get it working!

I added `runtimeClassName: nvidia` to the spec of the Deployments that need HW acceleration like this example:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: object-detection-app-1
spec:
  replicas: 1
  strategy:
    type: Recreate
  selector:
    matchLabels:
      app: object-detection-app-1
  template:
    metadata:
      labels:
        app: object-detection-app-1
    spec:
      runtimeClassName: nvidia
      containers:
        - name: object-detection-app-1
... 
```

But my other comments are still valid. ;-)

**Comment by muhammadnasr** [comment]
How are you Anders?

I got this email notification from you. Are you sure you removed me from
all github projects?

**Comment by tekn0ir** [comment]
Hi!
All good here! I hope you are well also, sir!

That one is a public repo, so maybe you checked the ”watch repo” so you will be updated on changes?…

//A

**Pull request closed without merging**